src/relayserver/HttpServer.ts

```diff
@@ -54,46 +46,52 @@ export class HttpServer {
 
   close (): void {
     console.log('Stopping relay worker...')
-    this.backend.stop()
+    this.relayService?.stop()
   }
 
-  // TODO: use this when changing to jsonrpc
-  async rootHandler (req: any, res: any): Promise<void> {
-    let status
-    try {
-      let res
-      // @ts-ignore
-      const func = this.backend[req.body.method]
-      if (func != null) {
-        res = await func.apply(this.backend, [req.body.params]) ?? { code: 200 }
-      } else {
-        // @ts-ignore
-        // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
-        throw Error(`Implementation of method ${req.body.params} not found on backend!`)
-      }
-      status = jsonrpc.success(req.body.id, res)
-    } catch (e) {
-      let stack = e.stack.toString()
-      // remove anything after 'rootHandler'
-      stack = stack.replace(/(rootHandler.*)[\s\S]*/, '$1')
-      status = jsonrpc.error(req.body.id, new jsonrpc.JsonRpcError(stack, -125))
+  pingHandler (req: Request, res: Response): void {
+    if (this.relayService == null) {
+      throw new Error('RelayServer not initialized')
     }
-    res.send(status)
-  }
-
-  pingHandler (req: any, res: any): void {
-    const pingResponse = this.backend.pingHandler(req.query.paymaster)
+    const paymaster = req.query.paymaster
+    if (!(paymaster == null || typeof paymaster === 'string')) {
+      throw new Error('Paymaster address is not a valid string')
+    }
+    const pingResponse = this.relayService.pingHandler(paymaster)
     res.send(pingResponse)
     console.log(`address ${pingResponse.relayWorkerAddress} sent. ready: ${pingResponse.ready}`)
   }
 
-  async relayHandler (req: any, res: any): Promise<void> {
+  async relayHandler (req: Request, res: Response): Promise<void> {
+    if (this.relayService == null) {
+      throw new Error('RelayServer not initialized')
+    }
     try {
-      const signedTx = await this.backend.createRelayTransaction(req.body)
+      const signedTx = await this.relayService.createRelayTransaction(req.body)
       res.send({ signedTx })
     } catch (e) {
-      res.send({ error: e.message })
-      console.log('tx failed:', e)
+      const error: string = e.message
+      res.send({ error })
+      this.logger.error(`tx failed: ${error}`)
+    }
+  }
+
+  async penalizeHandler (req: Request, res: Response): Promise<void> {
```

Comment:
The idea with penalize in the server is that it gets a transaction from the client, and does the checks - the client simply sends all transactions to the relayer.
the client doesn't know if its a repeated nonce or anything else.

Reply:
Changed that